x/campaign/types/total_supply.go

```diff
@@ -21,3 +26,42 @@ func UpdateTotalSupply(coins, updatedCoins sdk.Coins) sdk.Coins {
 
 	return append(updatedCoins, notUpdated...).Sort()
 }
+
+// ValidateTotalSupply checks whether the total supply for each denom is within the provided total supply range
+func ValidateTotalSupply(totalSupply sdk.Coins, supplyRange TotalSupplyRange) error {
+	if err := supplyRange.Validate(); err != nil {
+		return err
+	}
+
+	for _, coin := range totalSupply {
+		totalSupply := coin.Amount
+		if totalSupply.LT(supplyRange.MinTotalSupply) {
+			return fmt.Errorf(
+				"total supply for %s cannot be lower than minTotalSupply %s: %s",
+				coin.Denom,
+				supplyRange.MinTotalSupply.String(),
+				coin.Amount.String())
+		}
+		if totalSupply.GT(supplyRange.MaxTotalSupply) {
+			return fmt.Errorf(
+				"total supply for %s cannot be greater than maxTotalSupply %s: %s",
+				coin.Denom,
+				supplyRange.MaxTotalSupply.String(),
+				coin.Amount.String())
+		}
+	}
+	return nil
+}
+
+// Validate performs basic validation on an instance of TotalSupplyRange
+func (sr TotalSupplyRange) Validate() error {
```

Comment:
We should also have `MinTotalSupply` being positive in the check

Reply:
Since it became identical to `validateTotalSupplyRange()` I replaced it with a call to that function

Reply:
There is maybe the interface conversion in addition in `validateTotalSupplyRange`

Reply:
perhaps we could do the other way around? (use this function in `validateTotalSupplyRange`?)